src/ffi.rs

```diff
@@ -4,686 +4,461 @@
-/// written in sequence to the given buffer. If there is no value and
-/// eligibility the first 16 bytes will be zero.
+/// Computes the total balance of the given notes.
+///
+/// Expects as argument a fat pointer to a JSON string representing
+/// [types::BalanceArgs].
+///
+/// Will return a triplet (status, ptr, len) pointing to JSON string
+/// representing [types::BalanceResult].
 #[no_mangle]
-pub unsafe extern "C" fn get_stake(
-    sk_index: *const u64,
-    stake: *mut [u8; StakeInfo::SIZE],
-) -> u8 {
-    let s = unwrap_or_bail!(WALLET.get_stake(*sk_index)).to_bytes();
-    ptr::copy_nonoverlapping(&s[0], &mut (*stake)[0], s.len());
-    0
-}
+pub fn balance(args: i32, len: i32) -> i64 {
+    let types::BalanceArgs { notes, seed } = match utils::take_args(args, len) {
+        Some(a) => a,
+        None => return utils::fail(),
+    };
 
-struct FfiStore;
+    let seed = match utils::sanitize_seed(seed) {
+        Some(s) => s,
+        None => return utils::fail(),
+    };
+
+    let notes: Vec<Note> = match rkyv::from_bytes(&notes) {
+        Ok(n) => utils::sanitize_notes(n),
+        Err(_) => return utils::fail(),
+    };
 
-impl Store for FfiStore {
-    type Error = u8;
+    let mut keys = unsafe { [mem::zeroed(); MAX_KEY + 1] };
+    let mut values = Vec::with_capacity(notes.len());
+    let mut keys_len = 0;
+    let mut sum = 0u64;
+
+    'outer: for note in notes {
+        // we iterate all the available keys until one can successfully decrypt
+        // the note. if all fails, returns false
+        for idx in 0..=MAX_KEY {
+            if keys_len == idx {
+                keys[idx] = key::derive_vk(&seed, idx as u64);
+                keys_len += 1;
+            }
 
-    fn get_seed(&self) -> Result<[u8; 64], Self::Error> {
-        let mut seed = [0; 64];
-        unsafe {
-            let r = get_seed(&mut seed);
-            if r != 0 {
-                return Err(r);
+            if let Ok(v) = note.value(Some(&keys[idx])) {
+                values.push(v);
+                sum = sum.saturating_add(v);
+                continue 'outer;
             }
         }
-        Ok(seed)
+
+        return utils::fail();
     }
+
+    // the top 4 notes are the maximum value a transaction can have, given the
+    // circuit accepts up to 4 inputs
+    values.sort_by(|a, b| b.cmp(a));
+    let maximum = values.iter().take(4).sum::<u64>();
+
+    utils::into_ptr(types::BalanceResponse {
+        maximum,
+        value: sum,
+    })
 }
 
-const STCT_INPUT_SIZE: usize = Fee::SIZE
-    + Crossover::SIZE
-    + u64::SIZE
-    + JubJubScalar::SIZE
-    + BlsScalar::SIZE
-    + Signature::SIZE;
+/// Computes a serialized unproven transaction from the given arguments.
+///
+/// Expects as argument a fat pointer to a JSON string representing
+/// [types::ExecuteArgs].
+///
+/// Will return a triplet (status, ptr, len) pointing to JSON string
+/// representing [types::ExecuteResult].
```

Comment:
This is supposed to be `[types::ExecuteResponse]`

Reply:
Thanks for catching that; fixed!